pkg/test/framework/integration/framework_test.go

```diff
@@ -16,125 +16,239 @@ package integration
+		for _, child := range tst.children {
+			subTest := t.NewSubTest(child.name)
+			if tst.runChildrenParallel {
+				subTest.RunParallel(child.runInternal)
+			} else {
+				subTest.Run(child.runInternal)
+			}
 		}
 	}
+	tst.runEnd = time.Now()
+}
+
+func (tst *test) timeRange() timeRange {
+	return timeRange{
+		start: tst.runStart,
+		end:   tst.cleanupEnd,
+	}
 }
 
-func runCheck(ctx framework.TestContext, errors chan error, name string) {
-	currentTest := atomic.NewString("")
-	for _, c := range []string{"a", "b", "c"} {
-		c := c
-		ctx.NewSubTest(c).Run(func(ctx framework.TestContext) {
-			// Store the test name. We will check this in ConditionalCleanup to ensure we call finish cleanup before the next test runs
-			currentTest.Store(c)
-			ctx.ConditionalCleanup(func() {
-				time.Sleep(time.Millisecond * 100)
-				if ct := currentTest.Load(); ct != c {
-					errors <- fmt.Errorf("expected current test for %s to be %s but was %s", name, c, ct)
+func (tst *test) doCheck() error {
+	// Make sure the component was closed after the test's run method exited.
+	if tst.componentCloseStart.Before(tst.runEnd) {
+		return fmt.Errorf("test %s: componentCloseStart (%v) occurred before runEnd (%v)",
+			tst.name, tst.componentCloseStart, tst.runEnd)
+	}
+
+	// Make sure the component was closed before the cleanup for this test was performed.
+	if tst.cleanupStart.Before(tst.componentCloseEnd) {
+		return fmt.Errorf("test %s: closeStart (%v) occurred before componentCloseEnd (%v)",
+			tst.name, tst.cleanupStart, tst.componentCloseEnd)
+	}
+
+	// Now make sure children cleanup occurred after cleanup for this test.
+	for _, child := range tst.children {
+		if child.cleanupEnd.After(tst.cleanupStart) {
+			return fmt.Errorf("child %s cleanupEnd (%v) occurred after parent %s cleanupStart (%v)",
+				child.name, child.cleanupEnd, tst.name, tst.cleanupStart)
+		}
+	}
+
+	if !tst.runChildrenParallel {
+		// The children were run synchronously. Make sure they don't overlap in time.
+		for i := 0; i < len(tst.children); i++ {
+			ci := tst.children[i]
+			ciRange := ci.timeRange()
+			for j := i + 1; j < len(tst.children); j++ {
+				cj := tst.children[j]
+				cjRange := cj.timeRange()
+				if ciRange.overlaps(cjRange) {
+					return fmt.Errorf("test %s: child %s[%s] overlaps with child %s[%s]",
+						tst.name, ci.name, ciRange, cj.name, cjRange)
 				}
-			})
-			for _, st := range []string{"p1", "p2"} {
-				ctx.NewSubTest(st).
-					RunParallel(func(ctx framework.TestContext) {})
 			}
-		})
+		}
+	}
+
+	// Now check the children.
+	for _, child := range tst.children {
+		if err := child.doCheck(); err != nil {
+			return err
+		}
 	}
+	return nil
+}
+
+func (tst *test) componentClosed(*component) {
+	tst.componentCloseStart = time.Now()
+	doWork()
+	tst.componentCloseEnd = time.Now()
+}
+
+func (tst *test) cleanup() {
+	tst.cleanupStart = time.Now()
+	doWork()
+	tst.cleanupEnd = time.Now()
+}
+
+func doWork() {
+	time.Sleep(time.Millisecond * 50)
```

Comment:
```suggestion
	time.Sleep(time.Millisecond
```

Do we need full 50ms sleep? Seems it is sufficient testing with very small and we do not get a ~15s runtime

Reply:
Reduced it to 10 millis